File: app/src/main/java/hr/marin/weatherjava/presenters/SearchPresenter.java
package hr.marin.weatherjava.presenters;

import java.util.List;

import hr.marin.weatherjava.models.SearchResult;
import hr.marin.weatherjava.services.WeatherService;
import hr.marin.weatherjava.services.listeners.IWeatherServiceListener;

public class SearchPresenter {
    private View view;
    private WeatherService service;

    public SearchPresenter(final View view) {
        this.view = view;
        this.service = new WeatherService();
    }

    public void search(String name) {
        service.findLocationByName(name, new IWeatherServiceListener<List<SearchResult>>() {
            @Override
            public void onSuccess(List<SearchResult> searchResults) {
                view.onSearchSuccess(searchResults);
            }

            @Override
            public void onFailure(Throwable t) {
                view.onError();
            }
        });
    }

    public interface View {
        void onSearchSuccess(List<SearchResult> resultList);
        void onError();
    }
}

File: app/src/main/java/hr/marin/weatherkotlin/presenters/SearchPresenter.kt
package hr.marin.weatherkotlin.presenters


import hr.marin.weatherkotlin.models.SearchResult
import hr.marin.weatherkotlin.services.WeatherService
import hr.marin.weatherkotlin.services.listeners.IWeatherServiceListener

class SearchPresenter(val view: View) {

    private val service: WeatherService = WeatherService()

    fun search(name: String) {
        service.findLocationByName(name, object: IWeatherServiceListener<List<SearchResult>> {
            override fun onSuccess(response: List<SearchResult>) {
                view.onSearchSuccess(response)
            }

            override fun onFailure(t: Throwable) {
                view.onError()
            }
        })
    }

    interface View {
        fun onSearchSuccess(resultList: List<SearchResult>)
        fun onError()
    }
}
